decoder.go

```diff
@@ -159,124 +161,151 @@ func (d *Decoder) unmarshalField(fieldDst reflect.Value, prop *ole.VARIANT) erro
-				}
-				t, err := time.Parse("20060102150405.000000-0700", val)
-				if err != nil {
-					return err
-				}
-				fieldDst.Set(reflect.ValueOf(t))
-			default:
-				return fmt.Errorf("can't deserealize string into %s", t)
-			}
+		return smartUnmarshalString(dst, val)
+	default:
+		return errSimpleVariantsExceeded
+	}
+	return nil
+}
+
+func unmarshalSlice(fieldDst reflect.Value, safeArray *ole.SafeArrayConversion) error {
+	arr := safeArray.ToValueArray()
+	resultArr := reflect.MakeSlice(fieldDst.Type(), len(arr), len(arr))
+	for i, v := range arr {
+		s := resultArr.Index(i)
+		err := unmarshalSimpleValue(s, v)
+		if err != nil {
+			return fmt.Errorf("can't put %T into []%s", v, fieldDst.Type().Elem().Kind())
+		}
+	}
+	fieldDst.Set(resultArr)
+	return nil
+}
+
+func smartUnmarshalString(fieldDst reflect.Value, val string) error {
+	switch fieldDst.Kind() {
+	case reflect.String:
+		fieldDst.SetString(val)
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		iv, err := strconv.ParseInt(val, 10, 64)
+		if err != nil {
+			return err
+		}
+		fieldDst.SetInt(iv)
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		uv, err := strconv.ParseUint(val, 10, 64)
+		if err != nil {
+			return err
+		}
+		fieldDst.SetUint(uv)
+	case reflect.Struct:
+		switch t := fieldDst.Type(); t {
+		case timeType:
+			return unmarshalTime(fieldDst, val)
+		default:
+			return fmt.Errorf("can't deserialize string into struct %T", fieldDst.Interface())
 		}
 	default:
-		if fieldDst.Kind() == reflect.Slice {
-			switch fieldDst.Type().Elem().Kind() {
-			case reflect.String:
-				safeArray := prop.ToArray()
-				if safeArray != nil {
-					arr := safeArray.ToValueArray()
-					fArr := reflect.MakeSlice(fieldDst.Type(), len(arr), len(arr))
-					for i, v := range arr {
-						s := fArr.Index(i)
-						s.SetString(v.(string))
-					}
-					fieldDst.Set(fArr)
-				}
-			case reflect.Uint8:
-				safeArray := prop.ToArray()
-				if safeArray != nil {
-					arr := safeArray.ToValueArray()
-					fArr := reflect.MakeSlice(fieldDst.Type(), len(arr), len(arr))
-					for i, v := range arr {
-						s := fArr.Index(i)
-						s.SetUint(reflect.ValueOf(v).Uint())
-					}
-					fieldDst.Set(fArr)
-				}
-			default:
-				return fmt.Errorf("unsupported slice type (%T)", val)
-			}
-		} else {
-			// If we got nil value - handle it with magic config fields.
-			gotNilProp := reflect.TypeOf(prop.Value()) == nil
-			if gotNilProp && (isPtr || d.NonePtrZero) {
-				ptrNeedZero := isPtr && d.PtrNil
-				nonPtrAllowNil := !isPtr && d.NonePtrZero
-				if ptrNeedZero || nonPtrAllowNil {
-					fieldDstOrig.Set(reflect.Zero(fieldDstOrig.Type()))
-				}
-				return nil
-			}
-			return fmt.Errorf("unsupported type (%T)", val)
+		return fmt.Errorf("can't deserealize string into %s", fieldDst.Kind())
+	}
+	return nil
+}
+
+func unmarshalTime(fieldDst reflect.Value, val string) error {
+	if len(val) == 25 {
```

Comment:
I get that it was this way before, but if you know what is happening a comment for magic constants like 25 & 22 would be nice.

Reply:
Drastically improved